src/D2L.CodeStyle.Analyzers/Immutability/ImmutabilityAnalyzer.cs

```diff
@@ -173,6 +183,54 @@ SimpleNameSyntax syntax
 			}
 		}
 
+		private static void AnalyzeConditionalImmutability(

```

Comment:
Rather than work at the syntax level here, I would try switching to symbols as quickly as possible.

```csharp
if( !( ctx.SemanticModel.GetDeclaredSymbol( ctx.Node ) is IMethodSymbol symbol ) ) {
  return;
}

foreach( IParameterSymbol parameter in symbol.TypeParameters ) {
  if( !Attributes.Objects.OnlyIf.IsDefined( parameter ) ) {
    continue;
  }

  // diagnostic
}
```

LocalFunction and MethodDeclaration should both provide IMethodSymbol even though GetDeclaredSymbol doesn't have an overload for LocalFunction specifically. You may still need to work with the input syntax node directly for placing the diagnostic.

Reply:
`Attributes.Objects.OnlyIf.IsDefined` does work with strings, but it does so in a well-defined way, and is relatively idiomatic in our codebase.

Reply:
Why symbols? These actions run later, its nice to do syntax when reasonable because it shows up in the editor faster.

Reply:
> Attributes.Objects.OnlyIf.IsDefined does work with strings, but it does so in a well-defined way, and is relatively idiomatic in our codebase.

I've been slowly removing it because it's bad :) It is used in these analyzers though.

Reply:
My suggestion is still registered for syntax analysis, the symbols just tend to be easier to work with?

Reply:
Oh yes, I misread.

(Also to be fair: waiting for the semantic model (which is common) mostly negates the latency benefits of syntax analyzers)

Reply:
I'm playing with Owen's idea. The problem is that I still need to hold some syntax to get the location of where to put the error, and it's actually less intuitive that way.

Reply:
> (Also to be fair: waiting for the semantic model (which is common) mostly negates the latency benefits of syntax analyzers)

Which would be true if comparing the attribute by symbol anyway, since we'd be calling `GetDeclaredSymbol`/`GetSymbolInfo` on each type parameter in order to do the comparison

Reply:
> I've been slowly removing it because it's bad :) It is used in these analyzers though.

Yep, I think it would be reasonable to make them non-static and add some sort of wrapper type that gets them lazily from the compilation. Having them easily reference-able by name and not repeated everywhere is pretty nice. I definitely did what you were suggesting outside of the immutability analyzers.

Reply:
I'm good either way here

Reply:
> I'm playing with Owen's idea. The problem is that I still need to hold some syntax to get the location of where to put the error, and it's actually less intuitive that way.

To avoid having to manage "syncing" the input syntax and the symbol, you should be able to use `parameter.DeclarationSyntaxReferences[0].GetSynax()` or similar. I wouldn't be terribly worried about the hop back to syntax land in this manner being maybe unperformant (comparatively), since it only happens in the error case. In the general case, the analyzers should produce no diagnostics.

Reply:
I've updated with another commit, it still uses the syntax reference from before for this purpose.